_posts/2019-12-10-variational-gaussian-process/vgp.Rmd

```diff
@@ -0,0 +1,549 @@
+
+The second thing to take care of is choosing the initial index points. From experiments, the exact choices don't matter that much, as long as the data are sensibly covered. For instance, an alternative way we tried was to construct an empirical distribution ([tfd_empirical](https://rstudio.github.io/tfprobability/reference/tfd_empirical.html)) from the data, and then sample from it. Here instead, we just use an -- unnecessary, admittedly, given the availability of `sample` in R -- fancy way to select random observations from the training data:
+
+```{r}
+num_inducing_points <- 50
+
+sample_dist <- tfd_uniform(low = 1, high = nrow(train) + 1)
+sample_ids <- sample_dist %>%
+  tfd_sample(num_inducing_points) %>%
+  tf$cast(tf$int32) %>%
+  as.numeric()
+sampled_points <- train[sample_ids, 1:8]
+```
+
+
+One interesting point to note before we start training. Computation of the posterior predictive parameters involves a Cholesky decomposition, which could fail if, due to numerical issues, the covariance matrix is no longer positive definite. A sufficient action to take in our case is to do all computations using `tf$float64`:
+
+```{r}
+k_set_floatx("float64")
+```
+
+Now we define (for real, this time) and run the model.
+
+```{r}
+model <- keras_model_sequential() %>%
+  layer_dense(units = 8,
+              input_shape = 8,
+              use_bias = FALSE) %>%
+  layer_variational_gaussian_process(
+    num_inducing_points = num_inducing_points,
+    kernel_provider = RBFKernelFn(),
+    event_shape = 1,
+    inducing_index_points_initializer = initializer_constant(as.matrix(sampled_points)),
+    unconstrained_observation_noise_variance_initializer =
+      initializer_constant(array(0.1))
+  )
+
+# KL weight sums to one for one epoch
+kl_weight <- batch_size / nrow(train)
+
+# loss that implements the VGP algorithm
+loss <- function(y, rv_y)
+  rv_y$variational_loss(y, kl_weight = kl_weight)
+
+model %>% compile(optimizer = optimizer_adam(lr = 0.008),
+                  loss = loss,
+                  metrics = "mse")
+
+history <- model %>% fit(train_ds,
+                         epochs = 100,
+                         validation_data = test_ds)
+
+plot(history)
+```
+
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = ""}
+knitr::include_graphics("images/history.png")
+```
+
+Interestingly, higher numbers of _inducing points_ (we tried 100 and 200) did not have much impact on regression performance ^[Performance being used in the sense of "lower regression error", not running speed. As to the latter, there definitely is a (negative) relationship between number of inducing points and training speed.]. Neither does the exact choice of multiplication constants (`0.1` and `2`) applied to the trained kernel `Variables` (`_amplitude` and `_length_scale`) 
```

Comment:
```suggestion
Interestingly, higher numbers of _inducing points_ (we tried 100 and 200) did not have much impact on regression performance ^[Performance being used in the sense of "lower regression error", not running speed. As to the latter, there definitely is a (negative) relationship between number of inducing points and training speed.]. Nor does the exact choice of multiplication constants (`0.1` and `2`) applied to the trained kernel `Variables` (`_amplitude` and `_length_scale`) 
```

Reply:
I think this should be nor, since it's the second of two things you're discussing that _did not have_ impact. (I could be wrong).

Reply:
oh, got it! Yes it's the second thing; I probably have had a wrong idea about how "neither" is used.